conn_introspect.go

```diff
@@ -0,0 +1,95 @@
+package swarm
+
+import (
+	introspectpb "github.com/libp2p/go-libp2p-core/introspect/pb"
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+type connIntrospector struct {
+	c *Conn
+	s *Swarm
+}
+
+func (ci *connIntrospector) id() string {
+	return ci.c.id
+}
+
+func (ci *connIntrospector) remotePeerId() string {
+	return ci.c.RemotePeer().String()
+}
+
+func (ci *connIntrospector) endPoints() *introspectpb.EndpointPair {
+	return &introspectpb.EndpointPair{SrcMultiaddr: ci.c.LocalMultiaddr().String(), DstMultiaddr: ci.c.RemoteMultiaddr().String()}
+}
+
+func (ci *connIntrospector) role() introspectpb.Role {
+	if ci.c.stat.Direction == network.DirInbound {
+		return introspectpb.Role_RESPONDER
+	} else {
+		return introspectpb.Role_INITIATOR
+	}
+}
+
+// TODO Number of packets & instantaneous bandwidth ?
+// Should we have a separate "flow-metre" for a connection to prepare for a message oriented world ?
+func (ci *connIntrospector) traffic() *introspectpb.Traffic {
+	if ci.s.bwc != nil {
```

Comment:
In these cases, prefer short-circuiting to reduce one level of nesting/indentation.

```go
if ci.s.bwc == nil {
  return nil
}
<rest>
```